### PR TITLE
BUG in Series.interpolate: limit_area/limit_direction kwargs with method="pad"/"bfill" have no effect

### LINKED ISSUE #26796: limit_area and limit_direction do not have an effect when interpolation method is 'pad'
#### Code Sample, a copy-pastable example if possible

```python
In [1]: import pandas as pd                                                                                                                                                                                                        

In [2]: import numpy as np                                                                                                                                                                                                         

In [3]: s = pd.Series([ 
   ...:     np.nan, 
   ...:     1., np.nan, 
   ...:     2., np.nan, np.nan, 
   ...:     5., np.nan, np.nan, np.nan, 
   ...:     -1., np.nan, np.nan 
   ...: ])                                                                                                                                                                                                                         

In [4]: s.interpolate(method='pad', limit_area='inside')                                                                                                                                                                           
Out[4]: 
0     NaN
1     1.0
2     1.0
3     2.0
4     2.0
5     2.0
6     5.0
7     5.0
8     5.0
9     5.0
10   -1.0
11   -1.0
12   -1.0
dtype: float64

# For method='linear' the `limit_area` kwarg works as expected

In [5]: s.interpolate(method='linear', limit_area='inside')                                                                                                                                                                        
Out[5]: 
0     NaN
1     1.0
2     1.5
3     2.0
4     3.0
5     4.0
6     5.0
7     3.5
8     2.0
9     0.5
10   -1.0
11    NaN
12    NaN
dtype: float64


```
#### Problem description

The kwargs `limit_area` and `limit_direction` for `interpolate()`, introduce in https://github.com/pandas-dev/pandas/pull/16513 do not seem to have an effect when using the method `pad`. They work with other interpolation methods, e.g. `linear`.

There are two different pathways for interpolate depending on the selected method. 
https://github.com/pandas-dev/pandas/blob/b8ad9da4213be55c587db015c640d927aeb7feaf/pandas/core/internals/blocks.py#L1096-L1116

For pad, ffill and bfill `_interpolate_with_fill()` is used which calls `missing.interpolate_2d` which does not seem to recognize the keywords limit_direction and limit_area. They are silently ignored.

I might be able to fix that during https://github.com/pandas-dev/pandas/pull/25141, but maybe a fix needs more fundamental changes.

#### Expected Output

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: eaacefd0957bfc6735c8782ac2ba571d7131d250
python: 3.7.2.final.0
python-bits: 64
OS: Darwin
OS-release: 16.7.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: None
LOCALE: de_DE.UTF-8

pandas: 0.25.0.dev0+725.geaacefd09
pytest: 4.1.1
pip: 18.1
setuptools: 40.6.3
Cython: 0.29.2
numpy: 1.15.4
scipy: 1.2.0
pyarrow: 0.11.1
xarray: 0.11.0
IPython: 7.2.0
sphinx: 1.8.2
patsy: 0.5.1
dateutil: 2.7.5
pytz: 2018.9
blosc: None
bottleneck: 1.2.1
tables: 3.4.4
numexpr: 2.6.9
feather: None
matplotlib: 3.0.2
openpyxl: 2.5.12
xlrd: 1.2.0
xlwt: 1.3.0
xlsxwriter: 1.1.2
lxml.etree: 4.3.0
bs4: 4.7.1
html5lib: 1.0.1
sqlalchemy: 1.2.16
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: 0.2.0
fastparquet: 0.2.1
pandas_gbq: None
pandas_datareader: None
gcsfs: None


</details>

### PR DESCRIPTION
**Pull request by arw2019**
- [x] closes #26796
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Picking up #31048

### CODE REVIEW

File: pandas/core/missing.py [hunk 1]
```diff
@@ -542,14 +542,55 @@ def _cubicspline_interpolate(xi, yi, x, axis=0, bc_type="not-a-knot", extrapolat
...
 ):
     """
     Perform an actual interpolation of values, values will be make 2-d if
     needed fills inplace, returns the result.
     """
+
+    # `limit_area` is not supported by `pad_2d` and `backfill_2d`. Hence, the
+    # following code block does a recursive call and applies the interpolation
+    # and `limit_area` logic along a certain axis.
+    if limit_area is not None:
```
Comment (arw2019) [inline pandas/core/missing.py:558]: working on https://github.com/pandas-dev/pandas/pull/31048/files#r526326003

File: pandas/core/missing.py [hunk 2]
```diff
@@ -542,14 +542,55 @@ def _cubicspline_interpolate(xi, yi, x, axis=0, bc_type="not-a-knot", extrapolat
...
+
+        def func(values):
```
Comment (jreback) [inline pandas/core/missing.py:560]: can you make this a module level function (can be private is ok), give it a real name, with doc-strings & types.

**Review by jreback — CHANGES_REQUESTED** [review]
ok for 1.2. ping when comments and ready.

File: pandas/core/missing.py (hunk 2 above)
Comment (arw2019) [reply pandas/core/missing.py]: Done

**Review by jreback — APPROVED** [review]

File: pandas/core/missing.py [hunk 3]
```diff
@@ -528,16 +528,65 @@ def _cubicspline_interpolate(xi, yi, x, axis=0, bc_type="not-a-knot", extrapolat
     return P(x)
 
 
+def _interpolate_with_limit_area(
+    values, method: str, limit: Optional[int], limit_area: Optional[str]
+):
+    """
+    Apply interpolation and limit_area logic to values along a to-be-specified axis.
```
Comment (jreback) [inline pandas/core/missing.py:535]: ok can you add a full doc-string here (and return type)

**Review by jreback — CHANGES_REQUESTED** [review]
great doc-string comment; pin gon green.

File: pandas/core/missing.py (hunk 3 above)
Comment (arw2019) [reply pandas/core/missing.py:535]: Done

File: pandas/core/missing.py [hunk 4]
```diff
@@ -528,16 +528,99 @@ def _cubicspline_interpolate(xi, yi, x, axis=0, bc_type="not-a-knot", extrapolat
...
+        )
+
+        if limit_area == "inside":
+            invalid[first : last + 1] = False
+        elif limit_area == "outside":
+            invalid[:first] = invalid[last + 1 :] = False
+
+        values[invalid] = np.nan
+    else:
+        values = interpolate_2d(
```
Comment (jreback) [inline pandas/core/missing.py:575]: this is hit? (this is weird because its *called* from interpolate_2d itself).

Comment (arw2019) [reply pandas/core/missing.py:575]: refactored this away (leftover from original PR)

Comment (arw2019) [reply pandas/core/missing.py:575]: it was hit for one example but `values` wasn't changed so may as well just drop through

**Comment by jreback** [comment]
hmm, precommit is failing.

**Comment by jreback** [comment]
@simonjayhawkins i expect to merge this for 1.2 but not blocking rc

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @arw2019

**Pull request merged**